ncm-afsclt: remove out of place code for managing PAM and iptables.

ajf8 opened the pull request:
It is better serviced in the proper dedicated components.

Our local version has diverged a lot from the upstream, which hasn't
been updated in many years. This patch adds NoAction support and a few
other fixes/features, but it also throws a lot of code for managing PAM
and iptables rules.

I think those features may only have used by one organisation which no
longer uses quattor.

ned21 commented:
"Sync to our local version" isn't very descriptive. Perhaps: "Remove out of place code for managing PAM and iptables that is better serviced in the proper dedicated components" ?

stdweird commented on ncm-afsclt/src/main/pan/components/afsclt/schema.pan:
```diff
@@ -1,19 +1,17 @@
 
 declaration template components/afsclt/schema;
 
 include { 'quattor/schema' };
```
no `{}`, only include the parts you need

stdweird commented on ncm-afsclt/src/main/pan/components/afsclt/schema.pan:
```diff
@@ -1,19 +1,17 @@
+   include structure_component
+   
    "thiscell"   : string         # "AFS home cell"
+   "thesecells" ? list           # "Cell list to authenticate to"
```
list of what?

stdweird commented on ncm-afsclt/src/main/pan/components/afsclt/schema.pan:
```diff
@@ -1,19 +1,17 @@
+   
    "thiscell"   : string         # "AFS home cell"
+   "thesecells" ? list           # "Cell list to authenticate to"
    "suidcells"  ? list           # "List of suid cells"
```
list of what?

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -15,74 +13,40 @@ use strict;
 #*** In example, $OSname can be 'Linux' or 'Solaris'
 #*** It's equivalent to perl's variable $^O (it gives the OS in lower-case)
 use LC::Sysinfo;
 my $OSname=LC::Sysinfo::os()->name;
```
is this used somewhere? maybe you can use `CAF::Service::os_flavour`?

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -15,74 +13,40 @@ use strict;
-# prevent authconfig from trying to launch in X11 mode
-delete($ENV{"DISPLAY"});
-
 #*** The following gives the OS name using the LC perl modules.
```
don't hesitate to cleanup the funky comment `#***` -> `#` :smile:

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -103,7 +67,7 @@ sub Configure_Cell {
     close(THISCELL);
```
FileReader?

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -128,361 +92,46 @@ sub Configure_Cell {
-
-    if ( ! -e "$pam_systemauth") {
-      $self->warn("Cannot find $pam_systemauth - not touching it");
+  my $thesecells = "/usr/vice/etc/TheseCells";
```
4 space indent

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -128,361 +92,46 @@ sub Configure_Cell {
+  my $thesecells = "/usr/vice/etc/TheseCells";
+
+  if ($config->elementExists($mypath."/thesecells")) {
+    if( ! $NoAction ) {
```
FileWriter.

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -128,361 +92,46 @@ sub Configure_Cell {
+        return;
+      }
+      foreach my $cell ( @$cells ) {
+        print THESECELLS $cell." ";
```
like `join(" ", @$cells)`?

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -128,361 +92,46 @@ sub Configure_Cell {
+      }
+      print THESECELLS "\n";
+      close(THESECELLS);
+      $self->OK("Configured cell list for authentication $thesecells");
```
`OK`? no `info`?

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -128,361 +92,46 @@ sub Configure_Cell {
-      if (! $libpam_options_refresh ) {
-         $self->debug(1, "No $pam_screensaver and no special options for it, ignoring");
+  } elsif ( -f $thesecells ) {
+    if( ! $NoAction ) {
```
no negative conditions in if/else, just do 
```perl
if ($NoAction) {
  would remove
} else {
  unlink
}
```

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -128,361 +92,46 @@ sub Configure_Cell {
+  } elsif ( -f $thesecells ) {
+    if( ! $NoAction ) {
+      if ( unlink($thesecells) ) {
+        $self->OK("Removed cell list for authentication $thesecells");
```
info?

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -503,30 +152,20 @@ sub Configure_Cache {
     $new_cache = $config->getValue($mypath."/cachesize");  #new cache size
   } else {
-    $self->warn("cannot get CDB $mypath/cachesize, giving up");
+    $self->warn("cannot get CDB $mypath/cachesize - not setting cache size");
```
what does CDB mean?

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -503,30 +152,20 @@ sub Configure_Cache {
-    }
-  }
-
   my $fs_run_cache_string =`fs getcacheparms 2>/dev/null`;
```
CAF::Process (with keeps_state?) (and can you check the code for more backticks or `qx//` or `system`? there's no real point in making NoAction effort if you don't use the CAF tools

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -536,112 +175,64 @@ sub Configure_Cache {
-      $self->error("cannot handle non-/afs mount point $afsmount in $afs_cacheinfo: ".$t[0]);
-      return 0;
-    }
+  if (open(CI, '<'.$afs_cacheinfo)) {
```
FileReader

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -536,112 +175,64 @@ sub Configure_Cache {
-    if($automatic_cache) {
-      $new_cache = "AUTOMATIC";
+  # sanity check - don't allow cachesize bigger than 95% of a partition size
+  my @df=`df -k $file_cachemount 2>/dev/null`;
```
backticks.....

stdweird replied:
@kwaegema might be nice thing to add to lib-blockdevices? (determine size from mounted FS via `df -k`)

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -536,112 +175,64 @@ sub Configure_Cache {
-	$self->error("cannot open $afs_cacheinfo for writing: $!");
-	return 0;
+  if ($new_cache != $file_cache) {
+      if (! $NoAction) {
```
FileWriter

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -536,112 +175,64 @@ sub Configure_Cache {
-    }
+      if (! $NoAction) {
+	  my $fscommand = "fs setcachesize $new_cache";
+	  my $fsout=`$fscommand 2>&1`;
```
grrrrrr.....

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -654,50 +245,22 @@ sub Configure_Cache {
+    $master_cellservdb = $config->getValue($mypath."/cellservdb");
+    if ( $master_cellservdb =~ m/^\/\w/i) { # non-URI -> file
+      $master_cellservdb = "file://".$master_cellservdb;
+    } elsif ( $master_cellservdb !~ m/^(ftp|http|file)/i) { # known URIs
```
`https?`

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -754,7 +317,7 @@ sub update_afs_cells ( $$ ) {
     @todo = ();
     my @current = `fs listcell -numeric 2>&1`;
```
Process with keeps_state

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -781,18 +344,22 @@ sub update_afs_cells ( $$ ) {
+	if (! $NoAction) {
+	    $error = 0;
+	    foreach $cell (@todo) {
+		my $out=`fs newcell $cell @{$ipaddrs{$cell}} 2>&1`;
```
guess what...

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -801,197 +368,113 @@ sub update_afs_cells ( $$ ) {
+	  my $args = $config->getElement($path)->getTree();
+	  my $content;
+	  $content .= "$_:" . $args->{$_} . "\n" for sort keys %$args;
+	  my $fh = CAF::FileWriter->new("/etc/afsd.args", log => $self, backup => ".old");
```
`CAF::FileWriter` is NoAction safe

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -801,197 +368,113 @@ sub update_afs_cells ( $$ ) {
-  my @newtables;
-  my $return = 1;
-  my $found = 0;
+=head1 NAME
```
move this to the top of the module and it's respective methods. it is much better to keep the pod close to the code

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -801,197 +368,113 @@ sub update_afs_cells ( $$ ) {
-  $self->info("AFS cache config not changed");
-  return 1;
-}
+=item /software/components/afsclt/active : boolean
```
this is a component, no need to document this for each component

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -801,197 +368,113 @@ sub update_afs_cells ( $$ ) {
-  if ($OSname =~ /Solaris/) {
-    $self->warn("Somebody needs to write something to unconfigure PAM on Solaris"); #FIXME
-    return 0;
+=item /software/components/afsclt/thiscell : string
```
documentation of the schema is best done via annotations in the schema. so this could be
```pan
type ... {
    ...
    @{local AFS cell for this machine}
     "thiscell" ? string
```
put the `@{}` block before the attribute

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -801,197 +368,113 @@ sub update_afs_cells ( $$ ) {
+
+=head1 BUGS
+
+The "debug" and "suidcell" parameters are silently ignored for now.
```
add deprecation warnings in the schema
```pan
    "suidcell" ? string with { deprecated(0, "suidcell is silently ignored"); true;}
```

ajf8 replied:
I decided to just delete these unimplemented options instead.

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -801,197 +368,113 @@ sub update_afs_cells ( $$ ) {
+
+nlist of arguments to stash in /etc/afsd.args used by rc.afs.
+
+=back
```
delete this stuff, just make sure that jaroslav is in the authors in the pom.xml (and maybe add yourself as maintainer?)

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -801,197 +368,113 @@ sub update_afs_cells ( $$ ) {
-##########################################################################
-  my ($self,$config)=@_;
+1; #required for Perl modules
 
```
this should be the last (empty) line in the component.

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -801,197 +368,113 @@ sub update_afs_cells ( $$ ) {
+      if (! $NoAction) {
+	  my $args = $config->getElement($path)->getTree();
+	  my $content;
+	  $content .= "$_:" . $args->{$_} . "\n" for sort keys %$args;
```
why no `print $fh "$_:" . $args->{$_} . "\n" for sort keys %$args;`, the extra variable makes no sense

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -801,197 +368,113 @@ sub update_afs_cells ( $$ ) {
+	  my $fh = CAF::FileWriter->new("/etc/afsd.args", log => $self, backup => ".old");
+	  print $fh $content;
+	  $fh->close();
+	  $self->OK("Configured afsd.args with $content");
```
info?

stdweird commented:
@ajf8 some advise in cleaning up: the CAF::Filewriter/Editor/Reader and Process are NoAction safe (won't do anything if NoAction is set)
for `CAF::Process` there exists a `keeps_state => 1` option to add during init, that will run a command even if NoAction is set (this is usefull to run harmless stuff like list something, so the component can proceed as much as allowed during NoAction)
i also do not know which version of CCM this is targeting, but as of 15.6, you can do `$config->getTree(/some/path)`, which returns undef if paths doesn't exists, so no more $config->elemenetexists / config->getvalue constructions

stdweird commented:
@ajf8  and thanks for the effort!

ajf8 commented:
Thanks a lot for the comments @stdweird, seems this module needed some attention, so I've tried to go through it all and tidy things up in general as well as addressing your comments.

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,314 @@
-#***    In Solaris the fs executable at /usr/afsws/bin
-#*** so we put that directory first in the path.
-if ($OSname=~ /Solaris/) {$ENV{'PATH'}="/usr/afsws/bin:$ENV{'PATH'}"}
+my $localcelldb   = '/usr/vice/etc/CellServDB';
```
if these are constants, use uppercase (and even better, readonly)
```perl
use Readonly;
Readonly my $LOCALCELLDB => '/usr/vice/etc/CellServDB';
```

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,314 @@
-    } else {
-      $enabled = 'on'; #default
+    my $proc = CAF::Process->new( [ "fs", "getcacheparms" ], log => $self, keeps_state => 1 );
+    if ( !( $? >> 8 ) && $proc->output() =~ /AFS using \d+ of the cache's available (\d+) (\w+) byte blocks/ ) {
```
can't check the exitcode before running something
```perl
my $output = CAF::Process->new( [ "fs", "getcacheparms" ], log => $self, keeps_state => 1 )->output();
if (!( $? >> 8 ) && $output && $output =~ /AFS using \d+ of the cache's available (\d+) (\w+) byte blocks/ ) {
...
```

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,314 @@
+        $run_cache = $1;
+    }
+    else {
+        $self->info("Cannot determine current AFS cache size, changing only config file");
```
isn't this at least `warn`?

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,314 @@
+        $self->info("Cannot determine current AFS cache size, changing only config file");
+    }
+
+    my $afs_cacheinfo_fh = CAF::FileEditor->new( $afs_cacheinfo, log => $self );
```
use `CAF::FileReader` if you are not going to modify anything

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,314 @@
+    # sanity check - don't allow cachesize bigger than 95% of a partition size
+    $proc = CAF::Process->new( [ "df", "-k", $file_cachemount ], log => $self, keeps_state => 1 );
+    for ( split ( "\n", $proc->output() ) ) {
+        if (m{^(.*?)\s+(\d+)\s+(\d+)\s+(\d+)\s+(\d+%)\s+(.*)}) {
```
don't group the patterns you don' need. it makes it much harder to read/interpret

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,314 @@
+
+    # sanity check - don't allow cachesize bigger than 95% of a partition size
+    $proc = CAF::Process->new( [ "df", "-k", $file_cachemount ], log => $self, keeps_state => 1 );
+    for ( split ( "\n", $proc->output() ) ) {
```
better not use `$_` in perl, it really does not behave as you expect. this better becomes
```perl
foreach my $line split ( "\n", $proc->output() ) ) {
   if ($line =~ m{...}) {
...
```
(`$_` is good for oneliners like `grep {$_ eq 'something} @array`, but for eveything else, keep it readable and use a variable)

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,314 @@
+            $proc = CAF::Process->new( $fscommand, log => $self );
+            $proc->run();
+            if ( $? >> 8 ) {
+                $self->warn( "Problem changing running AFS cache size via \"" . join ( " ", @$fscommand ) . "\":\n" + $proc->output() );
```
`CAF::Process` has stringification, so instead of `join ( " ", @$fscommand )`, you can just `"$proc"`
`proc->output()` will run the command a second time; so better to `$output = $proc->output()` instead of `->run` (i also think `run` dumps output to stdout, which noone wants)

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,314 @@
-    } else {
-      $self->OK("changed running AFS cache $run_cache -> $new_cache (1K blocks)");
+    else {
+        LC::Check::file(
```
use `FileWriter`

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,314 @@
+    my ( $cell, %seen, %ipaddrs, @hosts, $todo, @todo, $error, $afsutil );
 
-    while (<FH>) {
+    my $localcelldb_fh = CAF::FileEditor->new( $localcelldb, log => $self );
```
`FileReader` if you are only reading

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,314 @@
+    $localcelldb_fh->close();
+    foreach $cell ( keys(%ipaddrs) ) {
+        @{ $ipaddrs{$cell} } = sort( @{ $ipaddrs{$cell} } );
+        $self->debug( 1, "CellServDB $cell -> @{$ipaddrs{$cell}}" );
```
does this print anythin sensible?

ajf8 replied:
It prints what was parsed out of the existing cell DB, I don't mind dropping it.

stdweird replied:
leave it, i was curioes what the stringification of an array did.

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,314 @@
-    if($? >> 8 ) {
-      $self->info("cannot read current AFS cell info via fs:\n".join('',@current));
-      return 1;
+    if ( $? >> 8 ) {
```
don't chekc exitcode before output. creating the `CAF::Process` instance does not run anything.

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,314 @@
     }
-    foreach (@current) {
+
+    foreach ( split ( /\n/, $proc->output() ) ) {
```
use `my $line` instead of relying on `$_`

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,314 @@
-        if ("@hosts" ne "@{$ipaddrs{$cell}}") {
-            push(@todo, $cell);
-            $self->Debug("cell info for $cell changed from @hosts to @{$ipaddrs{$cell}}");
+        if ( "@hosts" ne "@{$ipaddrs{$cell}}" ) {
```
does this work? what does it compare?

ajf8 replied:
Yeah, it is comparing the IP addresses in the current CellServDB and the new one to determine if IP addresses have changed.

I guess this is a simple way of comparing two arrays by using their string representation, it does seem to work OK.

From 5.10 onwards the ~~ "smartmatch operator" is probably better. As we have 5.10 on RHEL4, I suppose this is a possibility.

stdweird replied:
no it seems to do space separated join, so for IP addresses this is safe.
```
# perl -e '@hosts=qw(a b c);$b={x => [qw(abc)]}; print "@hosts @{$b->{x}}\n"'
a b c abc
```
learned something new :wink:

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,314 @@
+                $proc = CAF::Process->new( [ "fs", "newcell", $cell, @{ $ipaddrs{$cell} } ], log => $self );
+                $proc->run();
+                if ( $? >> 8 ) {
+                    $self->error( "Error while updating AFS cell info for $cell: " . $proc->output() );
```
will run the command a second time. there is also no need for the `NoAction` block, `CAF::Process` is safe to use

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,314 @@
+    my $path = "$mypath/afsd_args";
+    if ( $config->elementExists($path) ) {
+        my $args = $config->getElement($path)->getTree();
+        my $fh = CAF::FileWriter->new( "/etc/afsd.args", log => $self, backup => ".old" );
```
the `"/etc/afsd.args"` should be a Readonly / constant

stdweird commented:
@ajf8 creating `CAF::Process` instances does not do anything, you need to `->output()` or `->execute()`. (i think `->run()` does execute with everything  redirected to stdout (and you don't want command output to stdout)

but almost there :smile:

stdweird commented on ncm-afsclt/src/main/pan/components/afsclt/schema.pan:
```diff
@@ -24,18 +21,7 @@ type component_afsclt_entry = {
-   "verbose"    ? string
-   "debug"      ? boolean
-   
+   "afsd_args"              ? nlist # /etc/afsd.args values for rc.afs
```
can you be a bit more specific? like `string{}`?

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,306 @@
-#***    In Solaris the fs executable at /usr/afsws/bin
-#*** so we put that directory first in the path.
-if ($OSname=~ /Solaris/) {$ENV{'PATH'}="/usr/afsws/bin:$ENV{'PATH'}"}
+my $mypath = '/software/components/afsclt';
```
also a readonly? (in principal, you can use `$self->prefix()` for this value, but it's used quite a lot, so variable is fine)

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,306 @@
-      $self->Unconfigure_firewall($config);
-      $self->Unconfigure_Cell($config);
-  }
+    my ( $self, $config ) = @_;
```
remove the `Unconfigure` completely if it is not supported

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,306 @@
-      return;
+    my ( $self, $config ) = @_;
+    unless ( $config->elementExists("$mypath/thiscell") ) {
+        $self->error("Cannot get $mypath/thiscell");
```
make the message clearer that this is a problem in the profile, not woth the component or the system itself (as this is mandatory in the schema)

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,306 @@
-  return;
-}
+    my $thiscell_fh = CAF::FileWriter->new( $THISCELL, log => $self );
+    print $thiscell_fh $afscell . "\n";
```
maybe `"$afscell\n"`?

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,306 @@
+    # cache size is configured in a separate function
+    if ( $config->elementExists("$mypath/afsmount") ) {
+        my $new_afsmount = $config->getValue("$mypath/afsmount");
+        if ( $new_afsmount ne '/afs' ) {
```
if only one value is allowed, also enforce it in the schema?

ajf8 replied:
I had a closer look at this, it seems totally useless. This parameter is not used anywhere.

it wouldn't work for us anyway because our afs is not mounted to /afs.

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,306 @@
+    if ( $config->elementExists("$mypath/afsmount") ) {
+        my $new_afsmount = $config->getValue("$mypath/afsmount");
+        if ( $new_afsmount ne '/afs' ) {
+            $self->error("Cannot handle non-/afs AFS mount point $new_afsmount");
```
this is a profile error, make it clear in the message.

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,306 @@
-	    );
+    else {
+        $self->info("Cannot get CDB $mypath/cachesize - not setting cache size");
+        return;
```
if the exitcodes are menaingful, this is success and maybe `return 1`?

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,306 @@
+        $self->warn("Cannot determine current AFS cache size, changing only config file");
+    }
+
+    my $afs_cacheinfo_fh = CAF::FileEditor->new( $AFS_CACHEINFO, log => $self );
```
`CAF::FileReader`

ajf8 replied:
Unfortunately we are still using CAF 13.1.0 on our older builds, so FileReader is not available.

stdweird replied:
ok, can you add the  following:
1. an item with future milestaone that says to use FileReader in afsctl
2. add a `# TODO use FileReader`
3. do a `$fs->cancel()` after each `->new` (it's basically what filereader does)

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,306 @@
+            my $disk_cachesize = $1;
+            my $mount          = $2;
+            if ( $mount eq $file_cachemount && $new_cache > 0.95 * $disk_cachesize ) {
+                $self->error("Cache size cannot exceed 95% of its partition size. Not changing.");
```
add mount and disk_cachesize  to the message?

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,306 @@
-    } else {
-      $self->info("AFS cache size is unchanged ($new_cache K)");
+
+    if ($NoAction) {
```
the `NoAction` is only needed for the `update_afs_cells` (or is that method alread `NoAction` safe on itself?

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,306 @@
-    while (<FH>) {
-        if (/^>(.\S+)/) {
-            # new cell
+    my $localcelldb_fh = CAF::FileEditor->new( $LOCALCELLDB, log => $self );
```
`FileReader`

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,306 @@
-      $self->warn("cannot read local $localcelldb, giving up");
-      return 0;
-    }
+    my ($self) = @_;
```
i think this method is `NoAction` safe on it's own. no need for the `if` construct above

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,306 @@
             $ipaddrs{$cell} = [];
-        } elsif (/^(\S+)\s+\#\s*(\S+)/) {
+        }
+        elsif ($line =~ /^(\S+)\s+\#\s*(\S+)/) {
```
no need for the second group

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,306 @@
-        chomp($_);
-        next unless /^Cell\s+(\S+)\s+on hosts\s+(\S.+)\.$/;
+
+    foreach my $line ( split ( /\n/, $proc->output() ) ) {
```
just `$output`

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,306 @@
+
     # check new cells
-    foreach $cell (keys(%ipaddrs)) {
+    foreach $cell ( keys(%ipaddrs) ) {
```
`sort keys` to make it reproducable

stdweird commented:
@ajf8 we're at the minor remarks :+1: :smile:

ajf8 commented:
Yep nearly there! It's a much better component now. Thanks again.

stdweird commented:
@ajf8 one minor remark left (i was suprised the noaction supported was there :smile: )
wrt using this with old `CAF`, i'm unaware if there are any NoAction-related bugs in previous releases, so you better test the noaction part

stdweird commented on ncm-afsclt/src/main/perl/afsclt.pm:
```diff
@@ -2,997 +2,281 @@
-$EC=LC::Exception::Context->new->will_store_all;
-
+$EC  = LC::Exception::Context->new->will_store_all;
 $NCM::Component::afsclt::NoActionSupported = 1;
```
use
```perl
our $NoActionSupported = 1;
```

(i totally missed that before, i also find it "intersting" this was set before)

stdweird commented:
LGTM
@ajf8 can you confirm that this code actually works. nice work!

ajf8 commented:
Thanks :+1:  Yeah, I have been testing it. In NoAction mode and not, checking the files come out the same, and editing/deleting them to make sure updates are tested.
